### PR TITLE
sourceNamespace and template parameters can be placed togather in httpprot

### PR DESCRIPTION
**Pull request by hhhcj**
As descirbe in the #903 , the commit let sourceNamespace and template parameters can be placed togather. Template option in requestBuilder takes higher priority than sourceNamespace. As member in template is non-default value, it will overlap the same member in sourceNamespace.

### CODE REVIEW

File: pkg/filters/builder/requestbuilder.go [hunk 1]
```diff
@@ -106,16 +102,20 @@ func (rb *RequestBuilder) Inherit(previousGeneration filters.Filter) {
...
 // Handle builds request.
 func (rb *RequestBuilder) Handle(ctx *context.Context) (result string) {
+	p := protocols.Get(rb.spec.Protocol)
 	if rb.spec.SourceNamespace != "" {
 		ctx.CopyRequest(rb.spec.SourceNamespace)
-		return ""
+		if rb.spec.Template == "" {
+			return ""
+		}
+		p.SetRef(ctx.GetOutputRequest())
```
Comment (localvar) [inline pkg/filters/builder/requestbuilder.go:118]: this function will be called concurrently, so this will not work.

**Comment by localvar** [comment]
@hhhcj , Thanks for the PR.
The code it changes is one of our most complex parts, so we will need time to check it carefully.

**Comment by samanhappy** [comment]
@hhhcj, thanks great for your contribution, I hava some doubts about this PR

- it seems that the logic in `BuildRequestWithRef` is modifying the current active request but not to build a new request from it, so it may in conflict with the design intent of `RequestBuilder` 
- I remember what you need is to copy some headers from other responses, but you are copying from a request, am I missing something?
- we are considering to enhance the `RequestAdaptor` to modify the original request from other responses, what do you think of it?

**Comment by hhhcj** [comment]
> we are considering to enhance the RequestAdaptor to modify the original request from other responses, what do you think of it?

I agree with it. I prefer this firstly, and I think we can place template option to all of the filters. When [localvar](https://github.com/localvar) said 
> For RequestAdaptor, we currently just want to keep it simple and don't tend to add template support to it.

 at #903, I turn to modify the RequestBuilder.

I will make a pr with template support for RequestAdaptor later. @samanhappy

**Comment by localvar** [comment]
hi @hhhcj , I'm closing the PR as it cannot be accepted:
* `RequestBuilder` is used to build a new request, not modify an existing one.
* the body (payload) of a request could be a [stream](https://github.com/megaease/easegress/blob/main/doc/reference/stream.md), in this case, it is cannot be copied from one request to another.
* a request instance could be referenced in more than one namespace, that's, even `sourceNamespace` and target namespace is identical, copy the existing body to another request may break other references in other namespaces.

**Pull request closed without merging**